github: support enterprise instances

beyang opened the pull request:
Thanks for making a wonderful library!

This PR adds support for GitHub Enterprise instances. It looks at the AuthURL endpoint to infer whether the GitHub instance is github.com or an enterprise instance.

dghubble commented on github/login.go:
```diff
@@ -83,3 +92,19 @@ func validateResponse(user *github.User, resp *github.Response, err error) error
 }
+
+func githubClientFromAuthURL(authURL string, httpClient *http.Client) (*github.Client, error) {
+	if strings.HasPrefix(authURL, "https://github.com/") {
```
It worries me some we can't have the regular Github Client as the default (else) behavior. I see why you've got it this way since GithubEnterprise auth url's don't look likely to have any structure we could detect and leverage.

I think everyone using gologin with Github today should have an Endpoint.AuthURL starting with `https://github.com` (Github forbits http), but there may be some cases someone proxies access through another endpoint.

beyang replied:
How about adding a `func EnterpriseCallbackHandler(config *oauth2.Config, success, failure http.Handler)` method that is the same as `CallbackHandler`, but sets an enterprise API root URL?

dghubble replied:
I'd rather not. Exposing the function would make Github Enterprise a first-class use case. Then if Github's enterprise product were to make a breaking change, I'd be expected to accomodate that in gologin.

As is, this PR would just let Github login automatically best-effort allow GHE (undocumented). If it becomes problematic, it can be reverted.

dghubble replied:
Taking a step back, if a users wants a handler for GHE, it would be very much in the gologin style to just add convenience functions that chain [gologin/oauth2](https://github.com/dghubble/gologin/tree/master/oauth2) with a `EnterpriseCallbackHandler`. And that handler can go in their code or here.

It really comes down to whether GHE should be in-scope for gologin or left to users.

dghubble replied:
Alright, I think your `EnterpriseCallbackHandler` is a reasonable approach. It eliminates the risk this will break anyone using regular Github login and it will be intuitive - its clear and isolated and avoids any magic of having GHE handling in the Github handler. And gologin is somewhat insulated from GHE making drastic breakages since the go-github library would have to deal with them first.

Ignore my "I'd rather not" from earlier.

dghubble commented on github/login.go:
```diff
@@ -83,3 +92,19 @@ func validateResponse(user *github.User, resp *github.Response, err error) error
+		if err != nil {
+			return nil, fmt.Errorf("could not parse AuthURL %s", authURL)
+		}
+		baseURL.Path = ""
```
Why not just alter the baseURL.Path to be the required `/api/v3/`? Avoid the string concatenation. Pass `baseURL.String()` directly.

Also, the docs mention `Note that NewEnterpriseClient is a convenience helper only; its behavior is equivalent to using NewClient, followed by setting the BaseURL and UploadURL fields.` That may be more readable.

dghubble commented on github/login.go:
```diff
@@ -83,3 +92,19 @@ func validateResponse(user *github.User, resp *github.Response, err error) error
+		baseURL.Path = ""
+		baseURL.RawQuery = ""
+		baseURL.Fragment = ""
+		baseURLStr := strings.TrimSuffix(baseURL.String(), "/") + "/api/v3/"
```
This needs a command to hint at what's being done.

```
# convert authURL to GHE baseURL https://<mycompany>.github.com/api/v3/
```

dghubble commented on github/login.go:
```diff
@@ -57,8 +60,14 @@ func githubHandler(config *oauth2.Config, success, failure http.Handler) http.Ha
-		githubClient := github.NewClient(httpClient)
+		githubClient, err := githubClientFromAuthURL(config.Endpoint.AuthURL, httpClient)
+		if err != nil {
+			ctx = gologin.WithError(ctx, fmt.Errorf("could not parse AuthURL %s", config.Endpoint.AuthURL))
```
The error message at this stage should be about the call being made (we don't know its a parse error at this level and shouldn't assume). Also for consistency with the rest of this package, there's the `github:` prefix.

Perhaps `fmt.Errorf("github: failed to create github Client: %v", err)`

dghubble commented on github/login.go:
```diff
@@ -83,3 +92,19 @@ func validateResponse(user *github.User, resp *github.Response, err error) error
+	} else {
+		baseURL, err := url.Parse(authURL)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse AuthURL %s", authURL)
```
Let's use the github prefix and mention the Endpoint type since its what the user will need to correct.

`github: error parsing Endoint.AuthURL: %s`

dghubble commented on github/login.go:
```diff
@@ -83,3 +92,19 @@ func validateResponse(user *github.User, resp *github.Response, err error) error
 	return nil
 }
+
+func githubClientFromAuthURL(authURL string, httpClient *http.Client) (*github.Client, error) {
```
This should be table testable, you can compare the Client's BaseURL with an expected one.

dghubble commented:
Rebasing should solve the go get dependency failures.

beyang commented:
Addressed comments, PTAL

dghubble commented on github/login_test.go:
```diff
@@ -108,3 +109,23 @@ func TestValidateResponse(t *testing.T) {
+		authURL          string
+		expClientBaseURL string
+	}{
+		{authURL: "https://github.com/login/oauth/authorize/", expClientBaseURL: "https://api.github.com/"},
```
You don't have to repeat the field name if you don't want.

dghubble commented on github/login_test.go:
```diff
@@ -108,3 +109,23 @@ func TestValidateResponse(t *testing.T) {
+		{authURL: "http://github.mycompany.com/login/oauth/authorize", expClientBaseURL: "http://github.mycompany.com/api/v3/"},
+	} {
+		client, err := githubClientFromAuthURL(test.authURL, nil)
+		if err != nil {
```
assert.NotNil(err)

dghubble commented on github/login_test.go:
```diff
@@ -108,3 +109,23 @@ func TestValidateResponse(t *testing.T) {
 }
+
+func Test_githubClientFromAuthURL(t *testing.T) {
+	for _, test := range []struct {
```
While this is valid, can you define cases separately? (kinda a nitpick, but it matches other libraries)

```
cases := []struct{
    ...
}{
    ...
}
for _, c := range cases {
   ...
}
```

dghubble commented on github/login.go:
```diff
@@ -83,3 +92,19 @@ func validateResponse(user *github.User, resp *github.Response, err error) error
+func githubClientFromAuthURL(authURL string, httpClient *http.Client) (*github.Client, error) {
+	client := github.NewClient(httpClient)
+	if !strings.HasPrefix(authURL, "https://github.com/") {
+		// convert authURL to GHE baseURL https://<mycompany>.github.com/api/v3/
```
Example isn't right, GHE wouldn't be hosted under github.com

dghubble commented on github/login_test.go:
```diff
@@ -108,3 +109,23 @@ func TestValidateResponse(t *testing.T) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		if got, want := client.BaseURL.String(), test.expClientBaseURL; got != want {
```
gologin already makes use of testify assert, which you can use. assert.Equal(t, client.BaseURL.String(), c.expectedBaseURL)

beyang commented:
Comments addressed, PTAL. Given the addition of `EnterpriseCallbackHandler`, renamed `githubClientFromAuthURL` to `enterpriseGithubClientFromAuthURL` and updated the test. Also added a test for `EnterpriseCallbackHandler`.

dghubble commented:
Squashed and rebased as e2bffa01eb289b6ae0e7a8534e8a03be40432dd2 btw. Thanks for contributing!
